File: apps/web/src/lib/db/schema.ts
import { sqliteTable, text, integer, uniqueIndex } from "drizzle-orm/sqlite-core";

export const users = sqliteTable("users", {
  id: text("id").primaryKey(),
  username: text("username").unique().notNull(),
  displayName: text("display_name"),
  email: text("email").unique(),
  pendingEmail: text("pending_email"),
  emailVerified: integer("email_verified", { mode: "boolean" }).notNull(),
  passwordHash: text("password_hash"),
  avatarUrl: text("avatar_url"),
  bio: text("bio"),
  createdAt: integer("created_at", { mode: "timestamp" }).notNull(),
});

export const sessions = sqliteTable("sessions", {
  id: text("id").primaryKey(),
  userId: text("user_id").notNull().references(() => users.id),
  token: text("token").unique().notNull(),
  expiresAt: integer("expires_at", { mode: "timestamp" }).notNull(),
  createdAt: integer("created_at", { mode: "timestamp" }).notNull(),
});

export const oauthAccounts = sqliteTable("oauth_accounts", {
  id: text("id").primaryKey(),
  userId: text("user_id").notNull().references(() => users.id),
  provider: text("provider").notNull(),
  providerAccountId: text("provider_account_id").notNull(),
  createdAt: integer("created_at", { mode: "timestamp" }).notNull(),
}, (table) => ({
  providerAccountUnique: uniqueIndex("oauth_accounts_provider_account_unique").on(table.provider, table.providerAccountId),
  userProviderUnique: uniqueIndex("oauth_accounts_user_provider_unique").on(table.userId, table.provider),
}));

export const emailVerificationTokens = sqliteTable("email_verification_tokens", {
  id: text("id").primaryKey(),
  userId: text("user_id").notNull().references(() => users.id),
  tokenHash: text("token_hash").unique().notNull(),
  email: text("email").notNull(),
  expiresAt: integer("expires_at", { mode: "timestamp" }).notNull(),
  createdAt: integer("created_at", { mode: "timestamp" }).notNull(),
});

export const passkeys = sqliteTable("passkeys", {
  id: text("id").primaryKey(),
  userId: text("user_id").notNull().references(() => users.id),
  credentialId: text("credential_id").unique().notNull(),
  publicKey: text("public_key").notNull(),
  counter: integer("counter").notNull(),
  transports: text("transports"),
  deviceType: text("device_type"),
  backedUp: integer("backed_up", { mode: "boolean" }).notNull(),
  name: text("name"),
  createdAt: integer("created_at", { mode: "timestamp" }).notNull(),
  lastUsedAt: integer("last_used_at", { mode: "timestamp" }),
});

export const twoFactorSettings = sqliteTable("two_factor_settings", {
  userId: text("user_id").primaryKey().references(() => users.id),
  totpSecret: text("totp_secret"),
  recoveryCodes: text("recovery_codes"),
  enabled: integer("enabled", { mode: "boolean" }).notNull(),
  createdAt: integer("created_at", { mode: "timestamp" }).notNull(),
  updatedAt: integer("updated_at", { mode: "timestamp" }).notNull(),
});

File: apps/api-go/internal/auth/schema.go
// Package auth provides the Go data layer backing DiffAudit authentication:
// a SQLite schema compatible with the web app's drizzle schema, a store for
// users/sessions/oauth/verification records, password and token primitives,
// and request middleware for session checks.
package auth

import (
	"context"
	"database/sql"
	"fmt"
)

// Column names and constraints mirror apps/web/src/lib/db/schema.ts and the
// CREATE TABLE statements in apps/web/src/lib/db/index.ts (the runtime source
// of truth for existing databases). Timestamp columns are INTEGERs holding
// unix epoch seconds: the pinned drizzle-orm 0.45.2 serializes integer columns
// with mode "timestamp" as Math.floor(ms/1000).

const createUsersTable = `
CREATE TABLE IF NOT EXISTS users (
  id TEXT PRIMARY KEY,
  username TEXT UNIQUE NOT NULL,
  display_name TEXT,
  email TEXT UNIQUE,
  pending_email TEXT,
  email_verified INTEGER NOT NULL DEFAULT 0,
  password_hash TEXT,
  avatar_url TEXT,
  bio TEXT,
  created_at INTEGER NOT NULL
);`

const createSessionsTable = `
CREATE TABLE IF NOT EXISTS sessions (
  id TEXT PRIMARY KEY,
  user_id TEXT NOT NULL REFERENCES users(id),
  token TEXT UNIQUE NOT NULL,
  expires_at INTEGER NOT NULL,
  created_at INTEGER NOT NULL
);`

const createOAuthAccountsTable = `
CREATE TABLE IF NOT EXISTS oauth_accounts (
  id TEXT PRIMARY KEY,
  user_id TEXT NOT NULL REFERENCES users(id),
  provider TEXT NOT NULL,
  provider_account_id TEXT NOT NULL,
  created_at INTEGER NOT NULL,
  UNIQUE(provider, provider_account_id),
  UNIQUE(user_id, provider)
);`

const createEmailVerificationTokensTable = `
CREATE TABLE IF NOT EXISTS email_verification_tokens (
  id TEXT PRIMARY KEY,
  user_id TEXT NOT NULL REFERENCES users(id),
  token_hash TEXT UNIQUE NOT NULL,
  email TEXT NOT NULL,
  expires_at INTEGER NOT NULL,
  created_at INTEGER NOT NULL
);`

const createPasskeysTable = `
CREATE TABLE IF NOT EXISTS passkeys (
  id TEXT PRIMARY KEY,
  user_id TEXT NOT NULL REFERENCES users(id),
  credential_id TEXT UNIQUE NOT NULL,
  public_key TEXT NOT NULL,
  counter INTEGER NOT NULL,
  transports TEXT,
  device_type TEXT,
  backed_up INTEGER NOT NULL DEFAULT 0,
  name TEXT,
  created_at INTEGER NOT NULL,
  last_used_at INTEGER
);`

const createTwoFactorSettingsTable = `
CREATE TABLE IF NOT EXISTS two_factor_settings (
  user_id TEXT PRIMARY KEY REFERENCES users(id),
  totp_secret TEXT,
  recovery_codes TEXT,
  enabled INTEGER NOT NULL DEFAULT 0,
  created_at INTEGER NOT NULL,
  updated_at INTEGER NOT NULL
);`

// Migrate creates the auth tables when they do not exist yet. It is safe to
// run repeatedly; existing databases keep their current structure just like
// the web app's INIT_SQL bootstrap.
func Migrate(ctx context.Context, db *sql.DB) error {
	statements := []string{
		createUsersTable,
		createSessionsTable,
		createOAuthAccountsTable,
		createEmailVerificationTokensTable,
		createPasskeysTable,
		createTwoFactorSettingsTable,
	}
	for _, statement := range statements {
		if _, err := db.ExecContext(ctx, statement); err != nil {
			return fmt.Errorf("auth schema migrate: %w", err)
		}
	}
	return nil
}
